src/xen/xennet.c

```diff
@@ -99,60 +99,74 @@ struct xennet_dev {
     xennet_debug("%s: dev id %d", __func__, xd->if_id);
 
     u32 tx_id;
-    boolean again = false;
     char *node;
-    do {
-        s = xenstore_transaction_start(&tx_id);
-        if (!is_ok(s))
-            return s;
-
-        node = "vifname";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "vif%d", xd->if_id);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "rx-ring-ref";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->rx_ring_gntref);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "tx-ring-ref";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->tx_ring_gntref);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "request-rx-copy";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "feature-rx-notify";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "event-channel";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->evtchn);
-        if (!is_ok(s))
-            goto abort;
-
-        s = xenstore_transaction_end(tx_id, false);
-        if (!is_ok(s)) {
-            /* XXX check for EAGAIN */
-            goto abort;
+    int retries = XENNET_INFORM_BACKEND_RETRIES;
+
+  again:
+    node = "transaction start";
+    s = xenstore_transaction_start(&tx_id);
+    if (!is_ok(s))
+        return s;
+
+    node = "vifname";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "vif%d", xd->if_id);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "rx-ring-ref";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->rx_ring_gntref);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "tx-ring-ref";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->tx_ring_gntref);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "request-rx-copy";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "feature-rx-notify";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "event-channel";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->evtchn);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "transaction end";
+    s = xenstore_transaction_end(tx_id, false);
+    if (!is_ok(s)) {
+        value v = table_find(s, sym(errno));
+        if (v) {
+            if (!runtime_strcmp("EAGAIN", buffer_ref((buffer)v, 0))) {
+                deallocate_tuple(s);
+                if (retries-- == 0) {
+                    s = timm("result", "%s failed: transaction end returned EAGAIN after %d tries",
+                             __func__, XENNET_INFORM_BACKEND_RETRIES);
+                    goto abort;
+                }
+                goto again;
```

Comment:
I don't really know anything about this interface -- but is there any merit in pausing for a bit before retrying? What causes the EAGAINs?

Reply:
It's not clear to me if there is merit in pausing. Neither NetBSD nor Linux implementations pause before retrying.

Here the xennet driver is attempting to issue a xenstore "transaction" to the xenstore daemon. A transaction consists of multiple asynchronous (although since we're only using the xenstore for init, we're polling) writes to the store, delineated by transaction beginning and end messages. The xenstore daemon keeps a generation count of the database (or maybe some portion of it, I'm not sure). If a transaction begins and the generation count advances before the transaction end occurs, the transaction is aborted and an XS_ERROR message with a response of "EAGAIN" is sent to indicate to the caller that the transaction should be retried.

I suppose if there's a lot of contention on writes to the store then some kind of backoff delay might be in order, but I'm satisfied to play follow-the-leader here.

Reply:
ok gotcha